execution/scan/subquery.go

```diff
@@ -0,0 +1,142 @@
+type subqueryOperator struct {
+	next        model.VectorOperator
+	pool        *model.VectorPool
+	call        functionCall
+	mint        int64
+	maxt        int64
+	currentStep int64
+
+	funcExpr *parser.Call
+	subQuery *parser.SubqueryExpr
+
+	onceSeries sync.Once
+	series     []labels.Labels
+	acc        [][]sample
+}
+
+func NewSubqueryOperator(pool *model.VectorPool, next model.VectorOperator, opts *query.Options, funcExpr *parser.Call, subQuery *parser.SubqueryExpr) (model.VectorOperator, error) {
+	call, ok := rangeVectorFuncs[funcExpr.Func.Name]
+	if !ok {
+		return nil, parse.UnknownFunctionError(funcExpr.Func)
+	}
+	return &subqueryOperator{
+		next:        next,
+		call:        call,
+		pool:        pool,
+		funcExpr:    funcExpr,
+		subQuery:    subQuery,
+		mint:        opts.Start.UnixMilli(),
+		maxt:        opts.End.UnixMilli(),
+		currentStep: opts.Start.UnixMilli(),
+	}, nil
+}
+
+func (o *subqueryOperator) Explain() (me string, next []model.VectorOperator) {
+	return fmt.Sprintf("[*subqueryOperator] %v()", o.funcExpr.Func.Name), []model.VectorOperator{o.next}
+}
+
+func (o *subqueryOperator) GetPool() *model.VectorPool { return o.pool }
+
+func (o *subqueryOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+	if o.currentStep > o.maxt {
+		return nil, nil
+	}
+	if err := o.initSeries(ctx); err != nil {
+		return nil, err
+	}
+
+ACC:
+	for {
+		vectors, err := o.next.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+		if len(vectors) == 0 {
+			break ACC
+		}
+		for _, vector := range vectors {
+			for j, s := range vector.Samples {
+				o.acc[vector.SampleIDs[j]] = append(o.acc[vector.SampleIDs[j]], sample{T: vector.T, F: s})
+			}
+			o.next.GetPool().PutStepVector(vector)
+		}
+		o.next.GetPool().PutVectors(vectors)
+	}
+
+	res := o.pool.GetVectorBatch()
+	sv := o.pool.GetStepVector(o.currentStep)
+	for sampleId, rangeSamples := range o.acc {
+		f, h, ok := o.call(functionArgs{
+			Samples:     rangeSamples,
+			StepTime:    o.currentStep,
+			SelectRange: o.subQuery.Range.Milliseconds(),
+			Offset:      o.subQuery.Offset.Milliseconds(),
+		})
+		if ok {
+			if h != nil {
+				sv.AppendHistogram(o.pool, uint64(sampleId), h)
+			} else {
+				sv.AppendSample(o.pool, uint64(sampleId), f)
+			}
+		}
+	}
+	res = append(res, sv)
+
+	o.currentStep++
+	return res, nil
+}
+
+func (o *subqueryOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	if err := o.initSeries(ctx); err != nil {
+		return nil, err
+	}
+	return o.series, nil
+}
+
+func (o *subqueryOperator) initSeries(ctx context.Context) error {
+	var err error
+	o.onceSeries.Do(func() {
+		var series []labels.Labels
+		series, err = o.next.Series(ctx)
+		if err != nil {
+			return
+		}
+
+		o.series = make([]labels.Labels, len(series))
+		o.acc = make([][]sample, len(series))
+		for i, s := range series {
+			var b labels.ScratchBuilder
```

Comment:
We should move this out of the loop and call `Reset` on it here.

Reply:
Moved it out of the loop; `Reset` is already called in `DropMetricName` though so i left it out.